### PR TITLE
Prune zero-value reports

### PR DESCRIPTION
**Pull request by sdlyy**
Resolves L2B-2088

## What's changed

- No more zeroed values across repots
- Detailed controller is aware of potential data gaps due to lack of zeroed reports and fills the timeframes beforehand
- Data migration to get rid of zeroed values across reports
- insert filtering for zeroed reports

## How to test it?
Run staging dump locally and start local backend and frontend local variant. Data should be slim, charts will work as if no changes were made

### CODE REVIEW

File: packages/backend/src/core/reports/AggregatedReportUpdater.ts [hunk 1]
```diff
@@ -70,7 +71,9 @@ export class AggregatedReportUpdater {
       timestamp,
     )
 
-    await this.aggregatedReportRepository.addOrUpdateMany(aggregatedReports)
+    const nonZeroReports = aggregatedReports.filter((x) => x.usdValue > 0n)
+
+    await this.aggregatedReportRepository.addOrUpdateMany(nonZeroReports)
```
Comment (sz-piotr) [inline packages/backend/src/core/reports/AggregatedReportUpdater.ts:76]: What if the value was != 0 but changed to zero?

File: packages/backend/src/core/assets/CBVUpdater.ts [hunk 2]
```diff
@@ -105,7 +105,9 @@ export class CBVUpdater implements AssetUpdater {
     // TODO(radomski): This really needs to be refactored
     reports = filterOutNVMReports(reports)
 
-    await this.reportRepository.addOrUpdateMany(reports)
+    await this.reportRepository.addOrUpdateMany(
+      reports.filter((r) => r.amount > 0),
+    )
```
Comment (sz-piotr) [inline packages/backend/src/core/assets/CBVUpdater.ts:110]: Same concern: was != 0 but changed to zero.

File: packages/backend/src/core/assets/ArbitrumEBVUpdater.ts [hunk 3]
```diff
@@ -126,7 +126,10 @@ export class ArbitrumEBVUpdater implements AssetUpdater {
       this.getProjectId(),
       this.getChainId(),
     )
-    await this.reportRepository.addOrUpdateMany(reports)
+
+    await this.reportRepository.addOrUpdateMany(
+      reports.filter((r) => r.amount > 0),
```
Comment (sz-piotr) [inline packages/backend/src/core/assets/ArbitrumEBVUpdater.ts:131]: Same concern: was != 0 but changed to zero.

**Review by sz-piotr — CHANGES_REQUESTED** [review]
Maybe we should delete reports from a specific time and only then insert new ones (all in a transaction) instead of a simple adding of filtered ones?

File: packages/backend/src/peripherals/database/ReportRepository.ts [hunk 4]
```diff
@@ -63,25 +63,29 @@ export class ReportRepository extends BaseRepository {
     return rows.map(toRecord)
   }
 
-  async addOrUpdateMany(reports: ReportRecord[]) {
+  async addOrUpdateMany(
+    reports: ReportRecord[],
+    where: { timestamp: UnixTime; chainId: ChainId },
+  ) {
+    if (reports.length === 0) {
+      return 0
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/ReportRepository.ts:71]: we still need to remove records from the database in this case

File: packages/backend/src/peripherals/database/AggregatedReportRepository.ts [hunk 5]
```diff
@@ -112,23 +112,33 @@ export class AggregatedReportRepository extends BaseRepository {
     return row ? toRecord(row) : undefined
   }
 
-  async addOrUpdateMany(reports: AggregatedReportRecord[]) {
+  async addOrUpdateMany(
+    reports: AggregatedReportRecord[],
+    where: { timestamp: UnixTime },
+  ) {
+    if (reports.length === 0) {
+      return 0
```
Comment (sz-piotr) [inline packages/backend/src/peripherals/database/AggregatedReportRepository.ts:120]: we still need to remove records from the database in this case

**Review by sz-piotr — CHANGES_REQUESTED** [review]

File: packages/backend/src/peripherals/database/ReportRepository.ts (hunk 4 above)
Comment (sdlyy) [reply packages/backend/src/peripherals/database/ReportRepository.ts]: https://github.com/l2beat/l2beat/pull/1838/commits/65b872dba0346c18c980447d4e516c3ebbd9f1ed

File: packages/backend/src/peripherals/database/AggregatedReportRepository.ts (hunk 5 above)
Comment (sdlyy) [reply packages/backend/src/peripherals/database/AggregatedReportRepository.ts]: https://github.com/l2beat/l2beat/pull/1838/commits/65b872dba0346c18c980447d4e516c3ebbd9f1ed

**Comment by sdlyy** [comment]
Works locally - ignore unsynced data, please (it more or less proves that we actually fill the gaps ;p)


image

image

image

**Review by sz-piotr — APPROVED** [review]

**Review by antooni — CHANGES_REQUESTED** [review]
wait

**Comment by sdlyy** [comment]
Merging postponed for some time due to l2assets things

**Pull request closed without merging**